[SVG] Add ShapeType to LegacyRenderSVGShape

Constellation opened the pull request:
#### d6714182b3ab453ce39eee1ec18e3b5032dffec0
<pre>
[SVG] Add ShapeType to LegacyRenderSVGShape
<a href="https://bugs.webkit.org/show_bug.cgi?id=262724">https://bugs.webkit.org/show_bug.cgi?id=262724</a>
rdar://116538506

Reviewed by Simon Fraser.

This patch adds ShapeType to LegacyRenderSVGShape, which tells shape type of LegacyRenderSVGShape.
This enum will be used for efficient dispatch / behavior change for approximate repainting bounding box computation.

* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::isSingleDataLine const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
(WebCore::LegacyRenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
(WebCore::LegacyRenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::layout):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::shapeType const):

Canonical link: <a href="https://commits.webkit.org/268961@main">https://commits.webkit.org/268961@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/148e7c6414817d48a450b8ef0932cf9af589efda

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21137 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21497 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22198 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23012 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19650 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21377 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21694 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20890 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21361 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18332 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23865 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18227 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19162 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25423 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19308 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19364 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23359 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19896 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16916 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19178 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23464 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2620 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19757 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/Path.h:
```diff
@@ -92,6 +92,7 @@ class Path {
     std::optional<PathDataBezierCurve> singleBezierCurve() const;
 
     WEBCORE_EXPORT bool isEmpty() const;
+    bool isSingleDataLine() const;
```
Not clear what a "data line" is from outside of Path. Could this just be `isSingleLine`?

Constellation replied:
Sounds good. Changed.

smfr commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
```diff
@@ -61,6 +61,13 @@ void LegacyRenderSVGEllipse::updateShapeFromElement()
     if (m_radii.isEmpty())
         return;
 
+    if (m_radii.width() && m_radii.height()) {
```
`!m_radii.isEmpty()`

Constellation replied:
Changed.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -53,6 +53,14 @@ void LegacyRenderSVGPath::updateShapeFromElement()
+    else if (path().isSingleLine())
+        m_shapeType = ShapeType::Line;
+    else
+        m_shapeType = ShapeType::Path;
```
I do not know the purpose of this `ShapeType` and how much we should be accurate about it. But the `Path` can be any the types defined in `enum class ShapeType`. But surprisingly you handle the line only as special case.

Constellation replied:
The whole purpose of this `ShapeType` is optimization as described in the change comment, so we only handle the interesting one.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
```diff
@@ -42,6 +42,16 @@ class SVGGraphicsElement;
 class LegacyRenderSVGShape : public LegacyRenderSVGModelObject {
     WTF_MAKE_ISO_ALLOCATED(LegacyRenderSVGShape);
 public:
+    enum class ShapeType : uint8_t {
+        Empty,
+        Path,
+        Line,
+        Rectangle,
+        RoundedRectangle,
+        Ellipse,
+        Circle,
+    };
```
This enum is strange. `Path` can be any of the other ones. 

Also I wonder if it is better to add virtual methods instead. For example the RoundedRectangle can have this implementations:
```
class LegacyRenderSVGShape {
    virtual bool isRoundedRectangle() const { return m_path && m_path.isSingleRoundedRectangle(); }
}

class LegacyRenderSVGRect {
    bool isRoundedRectangle() const override { return rectElement().rx() > 0 || rectElement().ry() > 0; }
}
```

Constellation replied:
I don't think this is the right approach. Again, the whole purpose of this is optimization, and this is aligned to what LegacyRenderSVGShape is doing.
For example, you can create Rect by using path. But we don't use LegacyRenderSVGRect for that rect-shaping SVG path, right? The same thing can be applied here.

shallawa replied:
This type is in `RenderSVGShape` and it has to hold the right information about the shape that is going to be drawn regardless it is inherited by other classes or not.

Constellation replied:
How about changing the name to `ShapeTypeHint`? It is not planned to be arbitrary ShapeType for path. As we have different class for LegacyRenderSVGRect and LegacyRenderSVGPath, this is enum associated with these types + some optimization hint. Not something detecting the actual shape by more deep analysis.

Constellation replied:
We are not changing `LegacyRenderSVGPath` to `LegacyRenderSVGRect` when the path is actually drawing a Rect. It is still `LegacyRenderSVGPath`, but drawing the rect.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
```diff
@@ -42,6 +42,16 @@ class SVGGraphicsElement;
 class LegacyRenderSVGShape : public LegacyRenderSVGModelObject {
```
Is there a plan to port these changes to the LBSE classes: `RenderSVGShape`, `RenderSVGRect`, etc?

shallawa commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -421,6 +421,11 @@ bool Path::isEmpty() const
+bool Path::isSingleDataLine() const
+{
+    return !!singleDataLine();
+}
```
This will work only if the `Path` is a `PathSegment` or a `PathStream`. It will not work if it's `PathCG` even if the PathCG contains a single line.

Constellation replied:
Changed it to `definitelySingleLine`.

shallawa replied:
This is not what I meant. What I meant is `singleDataLine() `returns a `PathDataLine` which is a `PathSegment` which can be inside a `PathStream`. But it does not answer the question whether the path is going to draw a single line or not.

Constellation replied:
> This will work only if the Path is a PathSegment or a PathStream.

So, as you said, it works when this is `PathSegment` or `PathStream`, right?
And for `PathCG` for example, `singleDataLine` returns `std::nullopt`. So why doesn't `definitelySingleLine` work? If it returns `true`, then it is definitely a single line.
Or doesn't it work even with `PathSegment` or `PathStream`?

Constellation committed:
[SVG] Add ShapeType to LegacyRenderSVGShape
https://bugs.webkit.org/show_bug.cgi?id=262724
rdar://116538506

Reviewed by Simon Fraser.

This patch adds ShapeType to LegacyRenderSVGShape, which tells shape type of LegacyRenderSVGShape.
This enum will be used for efficient dispatch / behavior change for approximate repainting bounding box computation.

* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::isSingleDataLine const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
(WebCore::LegacyRenderSVGEllipse::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
(WebCore::LegacyRenderSVGRect::updateShapeFromElement):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::layout):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::shapeType const):

Canonical link: https://commits.webkit.org/268961@main
